README.md
# Fuzzy_logic_test_integer

This repository used for validation Fuzzy_logic library (integer variant). 
It is based on Visual Studio Code 1.66.1. 
compilator mingw64\\bin\\gcc.exe


CHANGELOG.md
# Changelog
All notable changes to Fuzzy_logic_test_integer will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), https://keepachangelog.com/ru/0.3.0/,
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
__________________________________________________________________________________________________________________________________________
## [Unreleased] 
### Added
- 
### Changed
- 
### Removed 
- 
__________________________________________________________________________________________________________________________________________
## [1.0.0] - 2022-04-20
### Added
- integer variables
- input data structure
- input scaling procedure
- output scaling procedure
### Changed
- cube algorithm
### Removed
- Gauss float fuzzification function
__________________________________________________________________________________________________________________________________________

src/main.c
//#include <bits/stdc++.h>

#include    <stdio.h>
#include    <stdbool.h>
#include    <stdlib.h>
#include    <stdint.h>
#include    <string.h>
#include    <math.h>
#include    "fuzzy_logic.h"

FILE *input_f;
FILE *output_f;

#define PI 3.1415926535897932384626433832795

/*************************** Fuzzy logic rules start ***************************************/
// Дистанция до линии в 2-х сантиметрах (limit +-127 = +-254см)
#define D_VERY_HIGH  (100)      // 2,0м
#define D_HIGH       (75)       // 1,5м
#define D_L1_CEN     (50)       // 1,0м
#define D_L1_TOP     (25)       // 50см
#define D_L1_BTN     (50)       // 1,0м
#define D_ZERO       (0)
#define D_LOW        (-D_HIGH)
#define D_VERY_LOW   (-D_VERY_HIGH)
#define D_Z_TOP      (1)        // 2см
#define D_Z_BTN      (10)       // 20см
#define D_R1_CEN     (-D_L1_CEN)
#define D_R1_TOP     (D_L1_TOP)
#define D_R1_BTN     (D_L1_BTN)


// Угол направления (theta) в градусах
#define T_VERY_HIGH  (115)
#define T_HIGH2      (20)
#define T_HIGH       (15)
#define T_ZERO       (0)
#define T_LOW        (-T_HIGH)
#define T_LOW2       (-T_HIGH2)
#define T_VERY_LOW   (-T_VERY_HIGH)

#define T_Z_TOP      (5)
#define T_Z_BTN      (10)
#define T_SIGMA      (10)

// Воздействие: угол направления колёс в градусах
#define TURN_H_LEFT     (25)
#define TURN_LEFT       (4)
#define NO_TURN         (0)
#define TURN_RIGHT      (-TURN_LEFT)
#define TURN_H_RIGHT    (-TURN_H_LEFT)

#define NULL_PARAM      (0)
#define NULL_OUT        (0)

// правила фуззификации (функции принадлежности)
//          name,     ffunc,    [n],  a,            b,              c,              next        
MAKE_FFUNC (d_zero,   trapecia, 0,    D_ZERO,       D_Z_TOP,        D_Z_BTN,        d_r1);     
MAKE_FFUNC (d_r1,     trapecia, 0,    D_R1_CEN,     D_R1_TOP,       D_R1_BTN,       d_r2);  
MAKE_FFUNC (d_r2,     low,      0,    D_VERY_LOW,   D_LOW,          NULL_PARAM,     d_l1);   
MAKE_FFUNC (d_l1,     trapecia, 0,    D_L1_CEN,     D_L1_TOP,       D_L1_BTN,       d_l2);   
MAKE_FFUNC (d_l2,     high,     0,    D_HIGH,       D_VERY_HIGH,    NULL_PARAM,     t_zero);   
MAKE_FFUNC (t_zero,   trapecia, 1,    T_ZERO,       T_Z_TOP,        T_Z_BTN,        t_l1);     
MAKE_FFUNC (t_l1,     triangle, 1,    T_LOW,        T_SIGMA,        NULL_PARAM,     t_l2);    
MAKE_FFUNC (t_l2,     low,      1,    T_LOW2,       T_LOW,          NULL_PARAM,     t_r1);   
MAKE_FFUNC (t_r1,     triangle, 1,    T_HIGH,       T_SIGMA,        NULL_PARAM,     t_r2);   
MAKE_FFUNC (t_r2,     high,     1,    T_HIGH,       T_HIGH2,        NULL_PARAM,     d_zero);   



// Правила нечёткой логики по данным эксперта
//        name,          A,          OPER,     B,        fin,       output,         next rule
MAKE_RULE (rule_01,      d_r2,       F_AND,    t_l2,     true,      NO_TURN,        rule_02);   // на большой дистанции угол въезда 25 градусов
MAKE_RULE (rule_02,      d_r2,       F_AND,    t_l1,     true,      TURN_LEFT,      rule_03);
MAKE_RULE (rule_03,      t_zero,     F_OR,     t_r1,     false,     NO_TURN,        rule_04);
MAKE_RULE (rule_04,      rule_03,    F_OR,     t_r2,     false,     NO_TURN,        rule_05);
MAKE_RULE (rule_05,      d_r2,       F_AND,    rule_04,  true,      TURN_H_LEFT,    rule_06);
MAKE_RULE (rule_06,      t_r2,       F_OR,     t_r1,     false,     NO_TURN,        rule_07);
MAKE_RULE (rule_07,      d_r1,       F_AND,    rule_06,  true,      TURN_H_LEFT,    rule_08);
MAKE_RULE (rule_08,      d_r1,       F_AND,    t_zero,   true,      TURN_LEFT,      rule_09);
MAKE_RULE (rule_09,      d_r1,       F_AND,    t_l1,     true,      NO_TURN,        rule_10);
MAKE_RULE (rule_10,      d_r1,       F_AND,    t_l2,     true,      TURN_RIGHT,     rule_11);
MAKE_RULE (rule_11,      d_zero,     F_AND,    t_l2,     true,      TURN_RIGHT,     rule_12);
MAKE_RULE (rule_12,      d_zero,     F_AND,    t_l1,     true,      NO_TURN,        rule_13);
MAKE_RULE (rule_13,      d_zero,     F_AND,    t_zero,   true,      NO_TURN,        rule_14);
MAKE_RULE (rule_14,      d_zero,     F_AND,    t_r1,     true,      NO_TURN,        rule_15);
MAKE_RULE (rule_15,      d_zero,     F_AND,    t_r2,     true,      TURN_LEFT,      rule_16);
MAKE_RULE (rule_16,      t_l2,       F_OR,     t_l1,     false,     NO_TURN,        rule_17);
MAKE_RULE (rule_17,      d_l1,       F_AND,    rule_16,  true,      TURN_H_RIGHT,   rule_18);
MAKE_RULE (rule_18,      d_l1,       F_AND,    t_zero,   true,      TURN_RIGHT,     rule_19);
MAKE_RULE (rule_19,      d_l1,       F_AND,    t_r1,     true,      NO_TURN,        rule_20);
MAKE_RULE (rule_20,      d_l1,       F_AND,    t_r2,     true,      TURN_LEFT,      rule_21);
MAKE_RULE (rule_21,      t_l2,       F_OR,     t_l1,     false,     NO_TURN,        rule_22);
MAKE_RULE (rule_22,      t_zero,     F_OR,     rule_21,  false,     NO_TURN,        rule_23);
MAKE_RULE (rule_23,      d_l2,       F_AND,    rule_22,  true,      TURN_H_RIGHT,   rule_24);
MAKE_RULE (rule_24,      d_l2,       F_AND,    t_r1,     true,      TURN_RIGHT,     rule_25);
MAKE_RULE (rule_25,      d_l2,       F_AND,    t_r2,     true,      NO_TURN,        rule_01);



#define MIN_0   (-250)      // min parameter #1 (delta in cm)
#define MAX_0   (250)       // max parameter #1

#define COUNT_0  (50)
#define MULT_0   ((MAX_0 - MIN_0) / COUNT_0)

#define MIN_1    (-30)      // min parameter #2 (course angle in degree)
#define MAX_1    (30)       // max parameter #2

#define COUNT_1    (60)
#define MULT_1    ((MAX_1 - MIN_1) / COUNT_1)


/*************************************************************************
 * @brief user scaling input 1 value to the limits +-127
 * input - distance in cm +-10000, limitation to +-254
 * @param in1 
 * @return int8_t 
*************************************************************************/
int8_t in0_scaling (int16_t in1)
{
    int16_t ret = in1 / 2;

    return (lim_s8 (ret));
}

/*************************************************************************
 * @brief user scaling input 2 value to the limits +-127
 * input - cource in degree +-180, limitation to +-127 degree
 * @param in2 
 * @return int8_t 
*************************************************************************/
int8_t in1_scaling (int16_t in2)
{
    int16_t ret = in2;

    return (lim_s8 (ret));
}

/*************************************************************************
 * @brief user scaling output value +-127 to the control value +-30 valve degree 
 * no scaling need
 * @param out       +-127 max
 * @return int16_t 
*************************************************************************/
int16_t out_scaling (int8_t out)
{
    return out;
}


int main()
{
    int8_t in[2], out, u;
    int16_t n, k, t;
    uint8_t z;
    (void)rule_01;  // avoid warnings
    (void)d_zero;

    fuzzy_param fuzzy = 
    {
        .in_array    = in,
        .start_ffunc = &d_zero,
        .start_rule  = &rule_01,
    };

    //input_f = fopen ("input.txt","r");
     // if (input_f == NULL)
    // {
    //     printf ("Error input!");   
    //     return (1);             
    // }
    output_f = fopen ("../output_int.txt","w");
    if (output_f == NULL)
    {
        printf ("Error output!");   
        return (1);             
    }

/// Validation fuzzification functions
/*
    /// cube
    printf ("Test fuzzy function cube with m=0, d=10\n");
    printf ("in\tout\n");
    for (k = -127; k < 128; k++)
    {
        z = cube (k, 0, 10, 0);
        printf("%d\t%d\n", k, z);
    }

    /// triangle
    printf ("Test fuzzy function triangle with m=0, d=100\n");
    printf ("in\tout\n");
    for (k = -127; k < 128; k++)
    {
        z = triangle (k, 0, 100, 0);
        printf("%d\t%d\n", k, z);
    }

    /// a_triangle
    printf ("Test fuzzy function a_triangle with m=20, d1=100, d2=50\n");
    printf ("in\tout\n");
    for (k = -127; k < 128; k++)
    {
        z = a_triangle (k, 20, 100, 50);
        printf("%d\t%d\n", k, z);
    }

    /// square
    printf ("Test fuzzy function square with m=-10, d1=50\n");
    printf ("in\tout\n");
    for (k = -127; k < 128; k++)
    {
        z = square (k, -10, 50, 0);
        printf("%d\t%d\n", k, z);
    }

    /// trapecia
    printf ("Test fuzzy function trapecia with m=-15, d1=10, d2=40\n");
    printf ("in\tout\n");
    for (k = -127; k < 128; k++)
    {
        z = trapecia (k, -15, 10, 40);
        printf("%d\t%d\n", k, z);
    }

    /// low
    printf ("Test fuzzy function low with min=-25, max=25\n");
    printf ("in\tout\n");
    for (k = -127; k < 128; k++)
    {
        z = low (k, -25, 25, 0);
        printf("%d\t%d\n", k, z);
    }

    /// high
    printf ("Test fuzzy function high with min=-5, max=55\n");
    printf ("in\tout\n");
    for (k = -127; k < 128; k++)
    {
        z = high (k, -5, 55, 0);
        printf("%d\t%d\n", k, z);
    }
*/

    printf ("Test fuzzy logic controller\n");
    fprintf (output_f, "\t");
    for (k = 0; k <= COUNT_0; k++)
    {
      // масштабированиение входных величин
        t = MIN_0 + (k * MULT_0);
        in[0] = in0_scaling (t);       // 2cm in the 1lsb
        fprintf (output_f, "%d\t", t);
    }
    fprintf (output_f, "\n");

    for (n = 0; n <= COUNT_1; n++)
    {
        t = MIN_1 + (n * MULT_1);
        in[1] = in1_scaling (t); 
        fprintf (output_f, "%d\t", t);

        for (k = 0; k <= COUNT_0; k++)
        {
            t = MIN_0 + (k * MULT_0);
            in[0] = in0_scaling (t);
            u = process_fuzzy_logic (&fuzzy);
            out = out_scaling (u);
            fprintf (output_f, "%d\t", out);
        }

        fprintf (output_f, "\n");
    }


    printf ("Test fuzzy logic controller end\n");

    // fclose (input_f);
    fclose (output_f);
}



src/fuzzy_logic.h
/*******************************************************************************
* \file     fuzzy_logic.h
* \author   Ilya Petrukhin (ilya.petrukhin@gmail.com)
* \brief 
* \version  2.0
* \date     2022-04-14
*******************************************************************************/

#ifndef _FUZZY_LOGIC_H_
#define _FUZZY_LOGIC_H_

/*******************************************************************************
* Rules to using fuzzy functions library
*******************************************************************************/
// 1. Define input values in the range +-127 (scale +-1.0f)
//  Need scaling input values to the range +-127
//  --------- Parameter #1 - input error --------------------------------
//  #define IN_VERY_HIGH  127             // 1.0f
//  #define IN_HIGH       32              // 0.25f
//  #define IN_LOW        -IN_HIGH        // -0.25f
//  #define IN_VERY_LOW   -IN_VERY_HIGH   // -1.0f
//  #define IN_ZERO       0               // 0.0f
//  #define IN_Z_TOP      25              // 0.2f
//  #define IN_Z_BTN      32              // 0.25f
//  --------- Parameter #2 - input delta error ---------------------------
//  #define DI_ZERO       0               // 0.0f
//  #define DI_ZERO_D     25              // 0.2f
//  #define DI_VERY_HIGH  63              // 0.5f
//  #define DI_VERY_LOW   -DI_VERY_HIGH   // -0.5f
   
// 2. Define output values in the range +-127 (after scaling if need) 
//  #define OUT_VERY_LOW  -127    // -100.0f
//  #define OUT_LOW       -13     // -10.0f
//  #define OUT_ZERO      0       // 0.0f
//  #define OUT_HIGH      13      // 10.0f
//  #define OUT_VERY_HIGH 127     // 100.0f

// 3. Define fuzzy input functions and register it into library
// each function have a link to the next, last function have a link to the first
// parameter "n" correspond to input parameters array number
// функции фуззификации 
//            name,     ffunc,    n,    a,            b,              c,              next        
//MAKE_FFUNC (mu_zero,  trapecia, 0,    IN_ZERO,      IN_Z_TOP,       IN_Z_BTN,       mu_low);      // нулевое значение отклонения ZERO
//MAKE_FFUNC (mu_low,   left,     0,    IN_VERY_LOW,  IN_VERY_HIGH,   NULL_PARAM,     mu_high);     // отклонение ниже нормы       LOW
//MAKE_FFUNC (mu_high,  right,    0,    IN_VERY_LOW,  IN_VERY_HIGH,   NULL_PARAM,     d_zero);      // отклонение выше нормы       HIGH
//MAKE_FFUNC (d_zero,   cube,     1,    DI_ZERO,      DI_ZERO_D,      NULL_PARAM,     d_low);       // нулевое значение скорости   ZERO
//MAKE_FFUNC (d_low,    left,     1,    DI_VERY_LOW,  DI_VERY_HIGH,   NULL_PARAM,     d_high);      // отклонение снижается        LOW
//MAKE_FFUNC (d_high,   right,    1,    DI_VERY_LOW,  DI_VERY_HIGH,   NULL_PARAM,     mu_zero);     // отклонение повышается       HIGH
   
// 4. Define fuzzy rules, maybe with some logic functions
// each rule have a link to the next, last rule have a link to the first
// complex rules have a some rows, and have links to the previous partially results
// Правила нечёткой логики по данным эксперта
//          name,          A,          OPER,     B,         fin,       output,        next rule
//MAKE_RULE (rule_zero,    mu_zero,    F_AND,    d_zero,    true,      OUT_ZERO,      rule_low);
//MAKE_RULE (rule_low,     mu_low,     F_OR,     d_zero,    true,      OUT_VERY_LOW,  rule_plow);
//MAKE_RULE (rule_plow,    mu_low,     F_OR,     d_low,     false,     OUT_ZERO,      rule_plow2);
//MAKE_RULE (rule_plow2,   rule_plow,  F_AND,    d_zero,    true,      OUT_LOW,       rule_high);
//MAKE_RULE (rule_high,    mu_high,    F_OR,     d_zero,    true,      OUT_VERY_HIGH, rule_zero);
  
// 5. Scale input values and put it into array and start process from the first rule
// put all values into fuzzy param structure
//  int8_t in[2];
//  fuzzy_param fuzzy = 
// {
//   .in_array    = in,
//   .start_ffunc = &mu_zero,
//   .start_rule  = &rule_zero,
// };
//
// input variables need to scale within +-127 range
//  in[0] = in0_scaling (in_err);
//  in[1] = in1_scaling (delta_err);
// 
// out variable also need to scale to output control value:
//  int8_t temp = process_fuzzy_logic (&fuzzy)
//  int16_t out = out_scaling (temp);
//
// ***************** end of the brief *****************************************
   
/// Fuzzy logic operators
typedef enum 
{
  F_AND  = 0,   ///< Fuzzy AND: a AND b = min(a,b)
  F_OR,         ///< Fuzzy OR: a OR b = max(a,b)
  F_NOT,        ///< Fuzzy NOT: NOT a = 1-a
  F_IMP,        ///< Fuzzy IMP: a -> b = b/a 
  F_A,          ///< Out is A
  F_B,          ///< Out is B
  F_FALSE       ///< Out is 0
} fuzzy_op;

typedef uint8_t (*fuzzy) (int8_t x,  int8_t p1, int8_t p2, int8_t p3);

/// Fuzzy logic function 
typedef struct 
{
  fuzzy     func;     ///< fuzzification function pointer
  uint8_t   xn;       ///< input parameter number
  int8_t    a;        ///< first function parameter
  int8_t    b;        ///< second function parameter
  int8_t    c;        ///< third function parameter
  uint8_t   y;        ///< fuzzification output value 
  void      *next;    ///< next fuzzy function pointer
} fuzzy_funct;  

#define MAKE_FFUNC(name, ffunc, x, a, b, c, next) \
  extern fuzzy_funct next;   \
	fuzzy_funct name = {ffunc, x, a, b, c, 0, &next}

/// Fuzzy rule control structure
typedef struct 
{
  uint8_t       *a;       ///< operand a pointer
  fuzzy_op      op;       ///< logic operator between a and b
  uint8_t       *b;       ///< operand b pointer
  bool          fin;      ///< flag final complex logic function
  uint8_t       y;        ///< fuzzy function result value
  int8_t        out;      ///< output fuzzy value
  void          *next;    ///< next rule pointer
} fuzzy_rules; 

/// Fuzzy parameters control structure
typedef struct 
{
  int8_t        *in_array;    ///< pointer input values array
  fuzzy_rules   *start_rule;  ///< pointer to the first fuzzy rule
  fuzzy_funct   *start_ffunc; ///< pointer to the first fuzzy function
} fuzzy_param;
   
#define MAKE_RULE(name, a, op, b, fin, out, next) \
  extern fuzzy_rules next;   \
	fuzzy_rules name = {&((a).y), op, &((b).y), fin, 0, out, &next}


/// Prototypes fuzzyfication input functions parameter x
uint8_t cube      (int8_t x, int8_t p1, int8_t p2, int8_t p3);  ///< Gauss cubic approximation, p1=M, p2=D_0.5 +-Delta 0.5 (n = 1 / D_0.5^3)
uint8_t triangle  (int8_t x, int8_t p1, int8_t p2, int8_t p3);  ///< symmetric triangle function p1=center, p1-p2=min, p1+p2=max _/\_
uint8_t a_triangle (int8_t x, int8_t p1, int8_t p2, int8_t p3); ///< asymmetric triangle function p1=center, p1-p2=min, p1+p3=max _/\_
uint8_t square    (int8_t x, int8_t p1, int8_t p2, int8_t p3);  ///< symmetric square function p1=center, p1-p2=min, p1+p2=max _|~|_
uint8_t trapecia  (int8_t x, int8_t p1, int8_t p2, int8_t p3);  ///< symmetric trapecial function p1=center, p1+p2=top, p1+p3=bottom _/~\_
uint8_t low       (int8_t x, int8_t p1, int8_t p2, int8_t p3);  ///< asymmetric low function p1=min p2=max ~\_
uint8_t high      (int8_t x, int8_t p1, int8_t p2, int8_t p3);  ///< asymmetric high function p1=min p2=max _/~

int8_t  in0_scaling (int16_t in1);  ///< user scaling input 1 value to the limits +-127
int8_t  in1_scaling (int16_t in2);  ///< user scaling input 2 value to the limits +-127
int16_t out_scaling (int8_t out);   ///< user scaling output value +-127 to the control value

int8_t process_fuzzy_logic (fuzzy_param *fuzzy);


inline static uint8_t   lim_u8  (int16_t x);
inline static int8_t    lim_s8  (int16_t x);

/*************************************************************************
 * \brief Limitation fuzzy result value by 0..1 e.g. 0..255
 *        for fuzzyfication functions
 * \param x         input
 * \return uint8_t  limitation result
*************************************************************************/
inline static uint8_t lim_u8 (int16_t x)
{
  if (x < 0)
  {
    x = 0;
  }
  else if (x > 255)
  {
    x = 255;
  }
  return (uint8_t)x;
}

/*************************************************************************
 * \brief Limitation in/out value by +-1 e.g. +-127
 *        for Fuzzyfication and Defuzzification function
 * \param x         input
 * \return int8_t   limitation output
*************************************************************************/
inline static int8_t lim_s8 (int16_t x)
{
  if (x < -127) // so
  {
    x = -127;
  }
  else if (x > 127)
  {
    x = 127;
  }
  return (int8_t)x;
}



#endif  // _FUZZY_LOGIC_H_


src/fuzzy_logic.c
/*******************************************************************************
* \file     fuzzy_logic.c
* \author   Ilya Petrukhin (ilya.petrukhin@gmail.com)
* \brief    This file provides code for fuzzy logic functions
* \version  2.0
* \date     2022-04-14
*******************************************************************************/
#include  <math.h>
#include  <stdio.h>
#include  <stdint.h>
#include  <stdbool.h>
#include  <string.h>
#include  "fuzzy_logic.h"

#define MIN(a,b)  ((a) < (b) ? (a) : (b))
#define MAX(a,b)  ((a) > (b) ? (a) : (b))


/*******************************************************************************
* кубическая аппроксимация гаусса, p1=M, p2=D_0.5
* \brief  Cubic approximation of Gauss function
* \param[in]  x   input value
* \param[in]  p1  Median
* \param[in]  p2  +-Delta 0.5 (n = 1 / D_0.5^3)
* \return         output value y = 1 / (1 + abs (n * (x - p1)^3)) 0-255
*******************************************************************************/
uint8_t cube (int8_t x, int8_t p1, int8_t p2, int8_t p3)
{
  int32_t d32;
  int32_t dx32;
  int32_t temp32;
  int16_t ret;
  
  (void)p3; 
  // y = D_0.5^3 / (D_0.5^3 + abs(x - p1)^3)
  if (p2 <0)
  {
    p2 = -p2;
  }
  d32 = p2;           // D_0.5
  d32 = (d32 * p2);   // D_0.5^2 * 127
  d32 = (d32 * p2);   // D_0.5^3 * 127^2

  if (x < p1)                         // abs(x-m)
  {
    dx32 = p1 - x;
  }
  else
  {
    dx32 = x - p1;                   
  }
  temp32 = (dx32 * dx32);     // (x-m)^2 * 127
  temp32 = (temp32 * dx32);   // (x-m)^3 * 127^2

  temp32 = temp32 + d32;      // d^3 + (x-m)^3

  if (temp32 != 0)
  {
    ret = d32 * 255 / temp32;  // d^3 / (d^3 + (x-m)^3)
  }
  else
  {
    ret = 255;
  }

  return lim_u8 (ret);
}


/*******************************************************************************
* симметричная треугольная функция p1=center, p1+p2=max, p1-p2=min _/\_
* \brief  Symmetric triangle function _/\_
* \param[in]  x   input value
* \param[in]  p1  Center
*                 p1 + p2 = max
*                 p1 - p2 = min
* \param[in]  p2  Delta 
* \return         output value
*******************************************************************************/
uint8_t triangle (int8_t x, int8_t p1, int8_t p2, int8_t p3)
{
  int16_t ret;
  int8_t min = p1 - p2;
  int8_t max = p1 + p2;
  
  (void)p3; 
  /// protection p2
  if (p2 == 0)
  {
    return 0;
  } 

  if ((x < min) || (x > max))
  {
    ret = 0;
  }
  else if (x < p1)
  {
    ret = ((int16_t)(x - min) * 255) / p2;
  }
  else
  {
    ret = ((int16_t)(max - x) * 255) / p2;
  }
  
  return lim_u8 (ret);
}

/*******************************************************************************
* Несимметричная треугольная функция p1=center, p1+p2=max, p1-p2=min _/\_
* \brief  Asymmetric triangle function _/\_
* \param[in]  x   input value
* \param[in]  p1  Center
*                 p1 + p3 = max
*                 p1 - p2 = min
* \param[in]  p2  Delta min
* \param[in]  p3  Delta max
* \return         output value
*******************************************************************************/
uint8_t a_triangle (int8_t x, int8_t p1, int8_t p2, int8_t p3)
{
  int16_t ret;
  int8_t min = p1 - p2;
  int8_t max = p1 + p3;

/// protection p2, p3
  if ((p2 == 0) || (p3 == 0))  
  {
    return 0;
  } 

  if ((x < min) || (x > max))
  {
    ret = 0;
  }
  else if (x < p1)
  {
    ret = ((int16_t)(x - min) * 255) / p2;
  }
  else
  {
    ret = ((int16_t)(max - x) * 255) / p3;
  }
  
  return lim_u8 (ret);
}

/*******************************************************************************
* симметричная прямоугольная функция p1=center, p1+p2=max, p1-p2=min _|~|_
* \brief  Symmetric square function _|~|_
* \param[in]  x   input value
* \param[in]  p1  Center
*                 p1 + p2 = max
*                 p1 - p2 = min
* \param[in]  p2  Delta 
* \return         output value
*******************************************************************************/
uint8_t square (int8_t x, int8_t p1, int8_t p2, int8_t p3)
{
  uint8_t ret = 0;

  (void)p3;  

  if ((x >= (p1 - p2)) && (x <= (p1 + p2)))
  {
    ret = 255;
  }
  return ret;
}

/*******************************************************************************
* симметричная трапециональная функция p1=center, p1+p2=max, p1-p2=min _/~\_
* основание p1+p3=max, p1-p3=min
* \brief  Symmetric trapecial function _/~\_
* \param[in]  x   input value
* \param[in]  p1  Center
* \param[in]  p2  p1 + p2 = top max, p1 - p2 = top min
* \param[in]  p3  p1 + p3 = base max, p1 - p3 = base min 
* \return         output value
*******************************************************************************/
uint8_t trapecia (int8_t x, int8_t p1, int8_t p2, int8_t p3)
{
  int16_t ret;
  int8_t min = p1 - p2;
  int8_t max = p1 + p2;
  int8_t min2 = p1 - p3;
  int8_t max2 = p1 + p3;
  
  if (x < min2 || x > max2)
  {
    ret = 0;
  }
  else if (x >= min && x <= max)
  {
    ret = 255;
  }
  else if (x < min)
  {
    if (min > min2)
    {
      ret = ((int16_t)(x - min2) * 255) / (min - min2);
    }
    else
    {
      ret = 255;
    }
  }
  else 
  {
    if (max2 > max)
    {
      ret = ((int16_t)(max2 - x) * 255) / (max2 - max);
    }
    else
    {
      ret = 255;
    }
  }
  
return lim_u8 (ret);
}

/*******************************************************************************
* несимметричная функция минимума p1=min p2=max ~\_
* \brief  Asymmetric minimum function ~\_
* \param[in]  x   input value
* \param[in]  p1  min
* \param[in]  p2  max 
* \return         output value
*******************************************************************************/
uint8_t low (int8_t x, int8_t p1, int8_t p2, int8_t p3)
{
  int16_t ret;
  int8_t temp;
  
  (void)p3;

  if (p1 > p2)        // swap p1, p2
  {
    temp = p1;
    p1 = p2;
    p2 = temp;
  }
  if (x < p1)
  {
    ret = 255;
  }
  else if (x > p2)
  {
    ret = 0;
  }
  else
  {
    if (p2 > p1)
    {
      ret = ((int16_t)(p2 - x) * 255) / (p2 - p1);
    }
    else
    {
      ret = 255;
    }
  }

return lim_u8 (ret);
}


/*******************************************************************************
* несимметричная функция максимума p1=min p2=max _/~
* \brief  Asymmetric maximum function _/~
* \param[in]  x   input value
* \param[in]  p1  min
* \param[in]  p2  max 
* \return         output value
*******************************************************************************/
uint8_t high (int8_t x, int8_t p1, int8_t p2, int8_t p3)
{
  int16_t ret;
  int8_t temp;

  (void)p3;  
  if (p1 > p2)  // swap p1, p2
  {
    temp = p1;
    p1 = p2;
    p2 = temp;
  }
  if (x < p1)
  {
    ret = 0;
  }
  else if (x > p2)
  {
    ret = 255;
  }
  else
  {
    if (p2 > p1)
    {
      ret = ((int16_t)(x - p1) * 255) / (p2 - p1);
    }
    else
    {
      ret = 255;
    }
  }
  
return lim_u8 (ret);
}




/*******************************************************************************
* Реализация нечеткого регулятора согласно правил fuzzy_param *fuzzy
* \brief Fuzzy logic controller by rules fuzzy_param *fuzzy
* \param[in] fuzzy->in_array[0] - error value (текущая ошибка)
* \param[in] fuzzy->in_array[1] - error diff (скорость изменения ошибки)
* \param[in] fuzzy->in_array[2] - error diff diff (option) 
*                                 (ускорение изменения ошибки (опция))
* \return Output control value 
*******************************************************************************/
int8_t process_fuzzy_logic (fuzzy_param *fuzzy)
{
  fuzzy_funct *f, *f1;
  fuzzy_rules *r, *r1;
  int16_t summ_alpha_c = 0;
  int16_t summ_alpha = 0;
  int16_t alpha, a, b, ret;
  int8_t *in_array;
  bool start;
  
  /// получить результаты функций фуззификации
  f1 = f = fuzzy->start_ffunc;
  in_array = fuzzy->in_array;
  start = true;
  while ((f != f1) || start)
  {
    if (f->func)
    {
      f->y = f->func (in_array[f->xn], f->a, f->b, f->c);
    }
    start = false;
    f = f->next;
  }
  
  /// цикл по правилам нечёткой логики
  r1 = r = fuzzy->start_rule;
  start = true;
  while ((r != r1) || start)
  {
    a = *(r->a);
    b = *(r->b);
    /// применяем логические операторы
    switch (r->op)
    {
    case F_AND:
      alpha = MIN (a, b);
      break;
      
    case F_OR:
      alpha = MAX (a, b);
      break;
      
    case F_NOT:
      alpha = 255 - a;
      break;
      
    case F_IMP:
      if (a == 0)
      {
        alpha = 255;
      }
      else
      {
        if (a)
        {
          alpha = (b * (int16_t)255) / a;
          alpha = lim_u8 (alpha);
        }
        else
        {
          alpha = 255;
        }
      }
      break;
      
    case F_A:
      alpha = a;
      break;
      
    case F_B:
      alpha = b;
      break;
      
    case F_FALSE:
    default:
      alpha = 0;
      break;
    }
    r->y = alpha;

    if (r->fin)  // если это конечное выражение
    {
      /// числитель и знаменатель для дискретного варианта 
      /// центроидного метода приведения к четкости
      summ_alpha_c += (alpha * (int16_t)r->out); // / 255;
      summ_alpha += alpha;
    }
    /// вычисляем следующее правило
    r = r->next;
    start = false;
  }

  /// вычисляем воздействие на объект управления
  if (summ_alpha == 0)
  {
    ret = 0;
  }
  else
  {
    ret = summ_alpha_c / summ_alpha;
  }
  return lim_s8 (ret);
}


